pkg/ebpf/map.go

```diff
@@ -70,13 +73,57 @@ func (m *Map) OpenOrCreate() error {
 	mapType := bpf.GetMapType(bpf.MapType(m.spec.Type))
 	m.spec.Flags = m.spec.Flags | bpf.GetPreAllocateMapFlags(mapType)
 
+	path := bpf.MapPath(m.spec.Name)
+
+	if m.spec.Pinning == ciliumebpf.PinByName {
+		mapDir := filepath.Dir(path)
+
+		if _, err := os.Stat(mapDir); os.IsNotExist(err) {
+			if err = os.MkdirAll(mapDir, 0755); err != nil {
+				return &os.PathError{
+					Op:   "Unable create map base directory",
+					Path: path,
+					Err:  err,
+				}
+			}
+		}
+	}
+
 	newMap, err := ciliumebpf.NewMapWithOptions(m.spec, opts)
 	if err != nil {
-		return fmt.Errorf("unable to create map: %w", err)
+		if !errors.Is(err, ciliumebpf.ErrMapIncompatible) {
+			return fmt.Errorf("unable to create map: %w", err)
+		}
+
+		// There already exists a pinned map but it has a different
+		// configuration (e.g different type, k/v size or flags).
+		// Try to delete and recreate it.
+
+		log.WithField("map", m.spec.Name).
+			Warnf("Removing map to allow for property upgrade (expect map data loss): %v", errors.Unwrap(err))
```

Comment:
Usually we use `.WithError()` to include an individual error, so that if a log shows up multiple times with different parameters, it is possible to correlate the common pieces in structured logging systems (eg the message is the same across all, but the error itself may vary). This can also help with grepping for an error as we can always tell that the `msg=` part is static text:

```suggestion
		log.WithField("map", m.spec.Name).
			WithError(err).Warn("Removing map to allow for property upgrade (expect map data loss)")
```

ref. https://github.com/cilium/cilium/issues/14156